lib/util_path.c

```diff
@@ -0,0 +1,206 @@
+ * - https://github.com/golang/go/blob/cfe2ab42/src/path/filepath/path.go#L88
+ *
+ * Copyright (c) 2009 The Go Authors. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <string.h>
+#include <stdlib.h>
+
+#include "xbps_api_impl.h"
+
+size_t
+xbps_path_clean(char *dst, size_t dstlen, const char *path)
+{
+	const char *p = path;
+	size_t dotdot = 0, len = 0;
+	bool rooted = *path == '/';
+
+#define append(c) do {        \
+	if (len+1 < dstlen)   \
+		dst[len] = c; \
+	len++;                \
+} while(0);
+
+	if (rooted) {
+		append('/');
+		p++;
+		dotdot++;
+	}
+
+	for (; *p;) {
+		switch (*p) {
+		case '/':
+			/* empty path element */
+			p++;
+			break;
+		case '.':
+			if (p[1] == '\0' || p[1] == '/') {
+				/* . element */
+				p++;
+			} else if (p[1] == '.' && (p[2] == '\0' || p[2] == '/')) {
+				p += 2;
+				/* .. element */
+				if (len > dotdot) {
+					/* can backtrack */
+					len--;
+					for (; len > dotdot && dst[len] != '/'; len--)
+						;
+				} else if (!rooted) {
+					/* cannot backtrack, but not rooted, append .. element. */
+					if (len > 0)
+						append('/');
+					append('.');
+					append('.');
+					dotdot = len;
+				}
+			}
+			break;
+		default:
+			if (len != (rooted ? 1 : 0))
+				append('/');
+			for (; *p && *p != '/'; p++)
+				append(*p);
+		}
+	}
+
+	/* Turn empty string into "." */
+	if (len == 0)
+		append('.');
+
+#undef append
+
+	dst[len < dstlen ? len : dstlen - 1] = '\0';
```

Comment:
Has the same semantics as strlcpy, snprintf etc.

Reply:
Fine. What about stating that in comment?

Reply:
Added the following to the header file for the functions this applies to:
```diff
+ * If the \a dst is too small to store the path it will be
+ * truncated and the returned length is the buffer size
+ * required to store the full path.
```

Reply:
I would prefer to error out rather than truncation. But ok.

Reply:
Hm yea, working with it, I don't really see a usecase for getting the length and the buffers will almost always be PATH_MAX.
Returning the required length makes sense for strlcpy/strlcat/snprintf to be able to allocate a buffer with the required size but here it doesn't make a lot of sense.

I'm think I'm going to change them to something like:

```
@return The length of the path or -1.
ssize_t xbps_path_append(char *dst, size_t len, char *suffix);
```

This would make error checks easier (avoids to have to do `>= sizeof buf`) and still allows to get the length in case you need it.